### PR TITLE
issue#17: Add logic for comparing objects from array

### PR DESCRIPTION
**Pull request by gorshkovayulia**
1. Move checking objects into separate class;
2. Add logic for comparing objects from array

### CODE REVIEW

File: core/src/main/java/io/elsci/assertreflectionequals/ErrorMessageBuilder.java [hunk 1]
```diff
@@ -56,28 +58,62 @@ public ErrorMessageBuilder addShallowDiff(Object expected, Object actual) {
...
      * @param actual actual array
-     * @param problemValues  contains mismatched values and corresponding index for array
+     * @param problemValues  contains mismatched values and their corresponding index in array
      */
     public ErrorMessageBuilder addArraysDiff(Object[] expected, Object[] actual, Map<Object, List<Object>> problemValues) {
-        this.addShallowDiff(Arrays.toString(expected), Arrays.toString(actual));
-
-        // For arrays only. We want to show which indices were different and values they had.
+        // Array may contain primitives or objects. We need to detect it before invoking 'addShallowDiff' method
+        // since if it's array with objects, we need firstly build an object array string
```
Comment (ctapobep) [inline core/src/main/java/io/elsci/assertreflectionequals/ErrorMessageBuilder.java:74]: Can you rephrase/elaborate on "object array string"? I didn't understand what this means.

File: core/src/main/java/io/elsci/assertreflectionequals/ErrorMessageBuilder.java [hunk 2]
```diff
@@ -56,28 +58,62 @@ public ErrorMessageBuilder addShallowDiff(Object expected, Object actual) {
...
+        if (expected == null && actual.length == 0 || actual == null && expected.length == 0) {
+            this.addShallowDiff(Arrays.toString(expected), Arrays.toString(actual));
+        } else if (expected == null || expected.length == 0) {
```
Comment (ctapobep) [inline core/src/main/java/io/elsci/assertreflectionequals/ErrorMessageBuilder.java:77]: Since this repeats.. Maybe add a method `isBlank()`? Similar to what they have in commons-collections

File: core/src/main/java/io/elsci/assertreflectionequals/ErrorMessageBuilder.java [hunk 3]
```diff
@@ -56,28 +58,62 @@ public ErrorMessageBuilder addShallowDiff(Object expected, Object actual) {
...
+            if (WRAPPER_TYPES.contains(actual[0].getClass())) {
+                this.addShallowDiff(Arrays.toString(expected), Arrays.toString(actual));
+            } else {
+                this.addShallowDiff(Arrays.toString(expected), toStringArrayWithObjects(actual));
+            }
+        } else if (actual == null || actual.length == 0) {
+            if (WRAPPER_TYPES.contains(expected[0].getClass())) {
```
Comment (ctapobep) [inline core/src/main/java/io/elsci/assertreflectionequals/ErrorMessageBuilder.java:84]: Actually arrays themselves have types. You don't have to access the 1st element to learn it:

```
      Class arrayClass = array.getClass();
      Class component = arrayClass.getComponentType();
```

Comment (gorshkovayulia) [reply core/src/main/java/io/elsci/assertreflectionequals/ErrorMessageBuilder.java:84]: Yes, I know about this way, but it doesnt' work here: the result will be always '_java.lang.Object_' because of calling 'getArrayWithValues' method and rebuilding array earlier

Comment (ctapobep) [reply core/src/main/java/io/elsci/assertreflectionequals/ErrorMessageBuilder.java:84]: Aaah, I see. You actually create an array of the right type earlier using Reflection: https://docs.oracle.com/javase/tutorial/reflect/special/arrayInstance.html

File: core/src/main/java/io/elsci/assertreflectionequals/ErrorMessageBuilder.java [hunk 4]
```diff
@@ -233,4 +237,11 @@ private StringBuilder toStringArrayWithObjects(Object[] array) {
         arrayWithObjects.append("]");
         return arrayWithObjects;
     }
+
+    /**
+     * Checks if array is null or empty
+     */
+    private boolean isBlank(Object[] object) {
```
Comment (ctapobep) [inline core/src/main/java/io/elsci/assertreflectionequals/ErrorMessageBuilder.java:244]: You can also turn it into `static`

**Comment by ctapobep** [comment]
+2
